Source/WebCore/rendering/style/RenderStyle.cpp

```diff
@@ -2638,41 +2651,39 @@ void RenderStyle::setColumnStylesFromPaginationMode(const Pagination::Mode& pagi
     }
 }
 
-void RenderStyle::deduplicateInheritedCustomProperties(const RenderStyle& other)
+void RenderStyle::deduplicateCustomProperties(const RenderStyle& other)
 {
-    auto& properties = const_cast<DataRef<StyleCustomPropertyData>&>(m_rareInheritedData->customProperties);
-    auto& otherProperties = other.m_rareInheritedData->customProperties;
-    if (properties.ptr() != otherProperties.ptr() && *properties == *otherProperties)
+    auto deduplicate = [&](auto& data, auto& otherData) {
+        auto& properties = const_cast<DataRef<StyleCustomPropertyData>&>(data->customProperties);
+        auto& otherProperties = otherData->customProperties;
+        if (properties.ptr() == otherProperties.ptr() || *properties != *otherProperties)
+            return;
         properties = otherProperties;
-}
-
-void RenderStyle::setInheritedCustomPropertyValue(const AtomString& name, Ref<CSSCustomPropertyValue>&& value)
-{
-    auto* existingValue = m_rareInheritedData->customProperties->values.get(name);
-    if (existingValue && existingValue->equals(value.get()))
-        return;
-    m_rareInheritedData.access().customProperties.access().setCustomPropertyValue(name, WTFMove(value));
-}
+        if (data == otherData)
+            data = otherData;
+    };
 
-void RenderStyle::setNonInheritedCustomPropertyValue(const AtomString& name, Ref<CSSCustomPropertyValue>&& value)
-{
-    auto* existingValue = m_rareNonInheritedData->customProperties->values.get(name);
-    if (existingValue && existingValue->equals(value.get()))
-        return;
-    m_rareNonInheritedData.access().customProperties.access().setCustomPropertyValue(name, WTFMove(value));
+    deduplicate(m_rareInheritedData, other.m_rareInheritedData);
+    deduplicate(m_rareNonInheritedData, other.m_rareNonInheritedData);
 }
 
-const CSSCustomPropertyValue* RenderStyle::customPropertyValue(const AtomString& name, const Style::CustomPropertyRegistry& registry) const
+void RenderStyle::setCustomPropertyValue(Ref<const CSSCustomPropertyValue>&& value, bool isInherited)
 {
-    if (auto* value = customPropertyValueWithoutResolvingInitial(name))
-        return value;
+    auto set = [&](auto& data) {
+        auto& name = value->name();
+        auto* existingValue = data->customProperties->values.get(name);
+        if (existingValue && existingValue->equals(value.get()))
+            return;
+        data.access().customProperties.access().setCustomPropertyValue(name, WTFMove(value));
+    };
 
-    // Alternatively we could just initialize the registered initial values to each RenderStyle.
-    auto* registered = registry.get(name);
-    return registered ? registered->initialValue.get() : nullptr;
+    if (isInherited)
+        set(m_rareInheritedData);
+    else
+        set(m_rareNonInheritedData);
```

Comment:
I'd use a ternary operator here, not a big deal obviously:
```suggestion
    set(isInherited ? m_rareInheritedData : m_rareNonInheritedData);
```

Reply:
That doesn't work since m_rareInheritedData and m_rareNonInheritedData have different types. The lambda here is a template really.